protocol_tests/connection/test_connection.py

```diff
@@ -4,13 +4,314 @@
+
+            # Sign it
+            response.sign(
+                signer=invite_verkey_b58,
+                secret=invite_sigkey
+            )
+            yield 'signed_response', conn, response
+
+            # Update connection info
+            info = request.get_connection_info()
+            conn.update(**info._asdict())
+
+        # If the agent wants to do a did exchange
+        elif msg['@type'] == (DidExchangeRequest.TYPE or DidExchangeRequest.ALT_TYPE):
+
+            # Validate their request
+            request = DidExchangeRequest(msg)
+            request.validate()
+
+            yield 'didexchange_request', request
+
+            # Get the did_doc~attach and verify the signature
+            attachment = request['did_doc~attach']
+            request.verify_signature(attachment['base64'], attachment['jws'])
+           
+            # Make the response
+            response = DidExchangeResponse.make(
+                request['@id'],
+                conn.did,
+                conn.verkey_b58,
+                config['endpoint'],
+                invite_sigkey,
+            )
+
+            yield 'didexchange_response', response
+
+            info = request.get_connection_info()
+            conn.update(**info._asdict())
+
+            # Send out the response and wait for the did exchange complete message
+            complete = (await conn.send_and_await_reply_async(
+                response, 
+                condition = lambda msg: msg.type == 'https://didcomm.org/didexchange/1.0/complete',
+                timeout=10
+            ))
+
+            # Double check that the thid and pthid are the proper values
+            assert complete["~thread"]["thid"] == request["@id"]
+            assert complete["~thread"]["pthid"] == invite["@id"]
+
+        else:
+            raise Exception("Expected Connection request or DID exchange request. Found: {}".format(msg['@type']))
+
+        yield 'connection_complete', conn
+            
+        # To finish up, let's go ahead and send out a trust ping to verify the connection works
+        await conn.send_and_await_reply_async(
+            {
+                '@type': 'https://didcomm.org/trust_ping/1.0/ping',
+                'response_requested': True
+            },
+            condition=lambda msg: msg.type == ('https://didcomm.org/trust_ping/1.0/ping_response' 
+                                                    or 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/trust_ping/1.0/ping_response'),
+            timeout=10,
+        )
+        
+        # Resgister a new handler for the out-of-band handshake reuse.
+        handler = HandshakeReuseHandler(invite['@id'])
+        conn.route_module(handler)
+
+        # Let's attempt to reuse the invitation here
+        await backchannel.out_of_band_v1_0_use_invitation(invite.to_url())
+
+        # Let's test the connection again with a trust ping
+        await conn.send_and_await_reply_async(
```

Comment:
Does the OOB protocol provide a way to exchange features that are supported.  I thought it did.  If yes, then you could do this only if they support the trust ping protocol.